### PR TITLE
SWATCH-648 SWATCH-723 SWATCH-724 Various fixes for account => orgId migration

### PR DESCRIPTION
**Pull request by mstead**
I'm sorry about the single PR that covers the 3 cards (subtasks) but as I was working on them, they were all sort of dependant on one another and were hard to untangle without conflicts.

1. https://issues.redhat.com/browse/SWATCH-648
    * Ignore account when not found during opt-in.
    * Remove unnecessary account checks when metering
2. https://issues.redhat.com/browse/SWATCH-723
    * Properly set account during hourly tally
    We were setting account_number on account calculations based on AccountInventoryService.account which may not exist if there is no account associated with an org. This patch will set it based on the first host instance that has a value set. This way the account number is in essance coming from the event that created the host.
3. https://issues.redhat.com/browse/SWATCH-724
    * Account number not required for awsUsageContext API

# How To Test

Please reach out if you have any issues with setting up for testing.

### Part 1

Generate some mock prometheus data without the ebs_account label and WITH external_organization set, and run the service. **Edit** the `bin/prometheus-mock-data.sh` script and remove the **ebs_account="...."** portion of the metric template.

```diff
-subscription_labels{_id="$CLUSTER_ID",billing_model="marketplace",ebs_account="$ACCOUNT",external_organization="$ORG_ID",support="Premium",billing_provider="$BILLING_PROVIDER",billing_marketplace_account="$MARKETPLACE_ACCOUNT",product="$PRODUCT"} 1.0 $TIME
+subscription_labels{_id="$CLUSTER_ID",billing_model="marketplace",external_organization="$ORG_ID",support="Premium",billing_provider="$BILLING_PROVIDER",billing_marketplace_account="$MARKETPLACE_ACCOUNT",product="$PRODUCT"} 1.0 $TIME

```

Run the script with a custom org. This should start a prometheus service for you.
```
ORG_ID=my-org ./bin/prometheus-mock-data.sh
```

Ensure your DB is fresh
```
dropdb -U rhsm-subscriptions rhsm-subscriptions && createdb -U rhsm-subscriptions rhsm-subscriptions
```

In a seperate terminal, launch the app (replace ${REPLACE} with act)
```
PROM_URL="http://localhost:9090/api/v1/" ./gradlew :bootRun
```

Run metering with your custom org from above.
```
curl -X POST -H "x-rh-swatch-synchronous-request: false" -H "x-rh-swatch-psk: placeholder" "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/rhosak?orgId=my-org&rangeInMinutes=120"
```

Check to make sure that metrics were found and that events were created successfully without an account number.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from events"
```

Run the hourly tally for this account. **Ensure to update the date range"
```
http :9000/hawtio/jolokia   type=exec   mbean=org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean   operation='tallyOrgByHourly(java.lang.String,java.lang.String,java.lang.String)'   arguments:='["my-org", "2022-12-13T00:00Z","2022-12-13T23:00Z"]'
```

Ensure that snapshots are produced (account number will not be set on the snapshots).
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from tally_snapshots"
```

### Part 2
**Revert the changes to the data generation script** and run it again with a custom org and account.
```
ORG_ID=my-org ACCOUNT=my-account ./bin/prometheus-mock-data.sh
```

Run metering with your custom org from above.
```
curl -X POST -H "x-rh-swatch-synchronous-request: false" -H "x-rh-swatch-psk: placeholder" "http://localhost:8000/api/rhsm-subscriptions/v1/internal/metering/rhosak?orgId=my-org&rangeInMinutes=120"
```

Check to make sure that metrics were found and that events were created successfully WITH an account number on NEW/UPDATED Events.
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from events"
```

Run the hourly tally for this account. **Ensure to update the date range"
```
http :9000/hawtio/jolokia   type=exec   mbean=org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean   operation='tallyOrgByHourly(java.lang.String,java.lang.String,java.lang.String)'   arguments:='["my-org", "2022-12-13T00:00Z","2022-12-13T23:00Z"]'
```

Ensure that snapshots are produced (account number WILL be set on NEW  snapshots).
```
psql -U rhsm-subscriptions rhsm-subscriptions -c "select * from tally_snapshots"
```

### Part 3
Create a custom product allow list with the following products.
```
MW02164HR
MW01882
```

Run the application configured against your custom allow list and production product/subscription services. **Be sure to replace the allow list path and truststore/passwords**

**!!!** The keystore/password for prod can be found in vault. **!!!**

```
PRODUCT_ALLOWLIST_RESOURCE_LOCATION=file:${PATH_TO_YOUR_ALLOW_LIST} PRODUCT_URL=https://product.api.redhat.com/svcrest/product/v3 PRODUCT_KEYSTORE=${YOUR_PATH}/keystore.jks PRODUCT_KEYSTORE_PASSWORD=${PROD_KEYSTORE_PASSWORD} RHSM_URL=https://api.rhsm.redhat.com/v1 RHSM_KEYSTORE=${YOUR_PATH}/keystore.jks RHSM_KEYSTORE_PASSWORD=${PROD_KEYSTORE_PASSWORD} SUBSCRIPTION_USE_STUB=false SUBSCRIPTION_KEYSTORE=${YOUR_PATH}/keystore.jks SUBSCRIPTION_KEYSTORE_PASSWORD=${PROD_KEYSTORE_PASSWORD}  SUBSCRIPTION_URL=https://subscription.api.redhat.com/svcrest/subscription/v5 DEV_MODE=true ENABLE_PAYG_SUBSCRIPTION_FORCE_SYNC=true PROM_URL="http://localhost:9090/api/v1/" ./gradlew :bootRun
```

Get the AWS usage context for org 13617887 (without account number) 
```
http GET :8000/api/rhsm-subscriptions/v1/internal/subscriptions/awsUsageContext orgId==13617887  productId==rhosak sla==Premium usage==Production date=="2022-12-09T19:00Z" awsAccountId==324151919439 x-rh-swatch-psk:placeholder
```

### CODE REVIEW

File: swatch-core/src/main/java/org/candlepin/subscriptions/db/AccountConfigRepository.java [hunk 1]
```diff
@@ -62,13 +63,17 @@ default Optional<AccountConfig> createOrUpdateAccountConfig(
...
       accountConfig.setOptInType(optInType);
       accountConfig.setCreated(current);
     }
-    accountConfig.setAccountNumber(account);
     accountConfig.setOrgId(orgId);
     accountConfig.setUpdated(current);
+
+    // Only set the account number if it is known.
+    if (StringUtils.hasText(account)) {
+      accountConfig.setAccountNumber(account);
```
[the remark this answers is not in the document]
Comment (mstead) [reply swatch-core/src/main/java/org/candlepin/subscriptions/db/AccountConfigRepository.java:74]: I'm not sure that we should do this unless we are removing the account completely from the table. We still want to ensure that an account is NOT represented by multiple orgs.

File: src/main/java/org/candlepin/subscriptions/security/OptInController.java [hunk 2]
```diff
@@ -104,12 +104,14 @@ public void optInByAccountNumber(String accountNumber, OptInType optInType) {
   // updates
   @Transactional(propagation = Propagation.REQUIRES_NEW)
   public String optInByOrgId(String orgId, OptInType optInType) {
-    if (orgConfigRepository.existsById(orgId)) {
+    if (accountConfigRepository.existsById(orgId)) {
```
[the remark this answers is not in the document]
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/security/OptInController.java]: DONE

File: src/main/java/org/candlepin/subscriptions/security/OptInController.java [hunk 3]
```diff
@@ -104,12 +104,14 @@ public void optInByAccountNumber(String accountNumber, OptInType optInType) {
...
       return accountConfigRepository.findAccountNumberByOrgId(orgId);
     }
-    String accountNumber = accountService.lookupAccountNumber(orgId);
-    log.info("Opting in account/orgId: {}/{}", accountNumber, orgId);
-    return performOptIn(accountNumber, orgId, optInType).getData().getAccount().getAccountNumber();
+
+    log.info("Opting in orgId={}", orgId);
+    // NOTE Passing null here should be cleaned up once account number
```
[the remark this answers is not in the document]
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/security/OptInController.java]: DONE

File: src/main/java/org/candlepin/subscriptions/metering/service/prometheus/PrometheusMeteringController.java [hunk 4]
```diff
@@ -230,20 +226,6 @@ Because our prometheus queries use a range vector (look back) of 1h, we need to
...
-      // Otherwise, we'd potentially end up in a state where a customer can't view the
-      // tally data.
-      throw new IllegalStateException("Refusing to persist event without accountNumber");
-    }
-    return accountNumberFromOptIn;
-  }
-
   private String ensureOptIn(String orgId) {
     try {
       return optInController.optInByOrgId(orgId, OptInType.PROMETHEUS);
```
[the remark this answers is not in the document]
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/metering/service/prometheus/PrometheusMeteringController.java]: DONE

**Review by kartikshahc — APPROVED** [review]

**Pull request merged**